File: crates/solobase-cloudflare/worker/project.ts
// Project resolution and feature gate logic for the TypeScript Cloudflare Worker.

import type { Env, ProjectConfig, ProjectAppConfig } from './types';

// ---------------------------------------------------------------------------
// Reserved subdomains — never resolve as a project
// ---------------------------------------------------------------------------

export const RESERVED_SUBDOMAINS = new Set([
  'app', 'admin', 'console', 'cloud', 'www', 'api', 'mail', 'smtp', 'ftp',
  'dashboard', 'billing', 'support', 'help', 'docs', 'blog', 'status',
  'staging', 'dev', 'test', 'demo', 'internal', 'private',
  'ns1', 'ns2', 'mx', 'autoconfig', 'autodiscover',
]);

// ---------------------------------------------------------------------------
// Project resolution
// ---------------------------------------------------------------------------

/**
 * Resolve a ProjectConfig from the request hostname.
 *
 * - Extracts the subdomain (first label before the first dot).
 * - localhost / 127.0.0.1 returns a hardcoded dev project with all features on.
 * - Otherwise looks up `project:{subdomain}:config` in the PROJECTS KV namespace.
 */
export async function resolveProject(
  hostname: string,
  env: Env,
): Promise<ProjectConfig | null> {
  // Strip port if present (e.g. "localhost:8787")
  const hostNoPort = hostname.split(':')[0];
  const subdomain = hostNoPort.split('.')[0];
  if (!subdomain) return null;

  // Reserved subdomains are the platform — never resolve as a project.
  const isDev = (env.ENVIRONMENT as string) === 'development';
  if (RESERVED_SUBDOMAINS.has(subdomain)) return null;
  if (isDev && (subdomain === 'localhost' || subdomain === '127' || subdomain === 'www')) {
    return {
      id: 'dev',
      subdomain: 'localhost',
      plan: 'hobby',
      db_binding: 'DB',
      config: allEnabledConfig(),
      blocks: [],
    };
  }

  // Look up project config in KV
  const kv = env.PROJECTS;
  if (!kv) return null;

  const key = `project:${subdomain}:config`;
  const raw = await kv.get(key, 'json');
  if (!raw) return null;

  return raw as ProjectConfig;
}

// ---------------------------------------------------------------------------
// Feature gates
// ---------------------------------------------------------------------------

/**
 * Check if a feature is enabled for the project.
 *
 * A feature is considered enabled when its field is present and truthy:
 * - `{}`, `true`, `{...}` -> enabled
 * - `undefined`, `null`, `false` -> disabled
 *
 * System and Profile are always enabled and do not need a feature gate.
 */
export function isFeatureEnabled(
  config: ProjectAppConfig,
  feature: string,
): boolean {
  switch (feature) {
    case 'system':
    case 'profile':
      return true;
    case 'auth':
      return isEnabled(config.auth);
    case 'admin':
      return isEnabled(config.admin);
    case 'files':
      return isEnabled(config.files);
    case 'products':
      return isEnabled(config.products);
    case 'projects':
      return isEnabled(config.projects);
    case 'legalpages':
      return isEnabled(config.legalpages);
    case 'userportal':
      return isEnabled(config.userportal);
    default:
      return false;
  }
}

function isEnabled(val: unknown): boolean {
  if (val === undefined || val === null || val === false) return false;
  return true;
}

// ---------------------------------------------------------------------------
// D1 binding resolution
// ---------------------------------------------------------------------------

/**
 * Get the D1Database binding for a project.
 *
 * Uses the project's `db_binding` field if set (e.g. `"DB_myapp"`),
 * otherwise falls back to the shared `DB` binding.
 */
export function getD1ForProject(env: Env, project: ProjectConfig): D1Database {
  if (project.db_binding) {
    const db = env[project.db_binding] as D1Database | undefined;
    if (db) return db;
  }
  return env.DB;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Create a ProjectAppConfig with all features enabled (dev / new project default). */
function allEnabledConfig(): ProjectAppConfig {
  return {
    version: 0,
    auth: {},
    admin: {},
    files: {},
    products: {},
    projects: {},
    legalpages: {},
    userportal: {},
  };
}

File: crates/solobase-cloudflare/src/project.rs
//! Project configuration, resolution, and plan limits.
//!
//! Projects are the multi-tenant units in Solobase Cloud. Each project
//! maps to a subdomain ({project}.solobase.dev) and has its own D1 database.
//!
//! Project data is stored in the platform D1 database (`projects` table).

use serde::{Deserialize, Serialize};
use worker::D1Database;

/// Reserved subdomains that cannot be used as project names.
pub const RESERVED_SUBDOMAINS: &[&str] = &[
    "admin", "api", "app", "auth", "billing", "blog", "cdn", "cloud",
    "console", "dashboard", "dev", "docs", "help", "internal", "login",
    "mail", "manage", "platform", "settings", "staging", "status",
    "support", "test", "www",
];

/// Check if a subdomain is reserved.
pub fn is_reserved_subdomain(subdomain: &str) -> bool {
    RESERVED_SUBDOMAINS.contains(&subdomain.to_lowercase().as_str())
}

/// Validate subdomain format: lowercase alphanumeric + hyphens, 1-63 chars,
/// must start/end with alphanumeric. Prevents path traversal in CF API URLs.
pub fn is_valid_subdomain(subdomain: &str) -> bool {
    if subdomain.is_empty() || subdomain.len() > 63 {
        return false;
    }
    let bytes = subdomain.as_bytes();
    if !bytes[0].is_ascii_alphanumeric() || !bytes[bytes.len() - 1].is_ascii_alphanumeric() {
        return false;
    }
    bytes.iter().all(|&b| b.is_ascii_lowercase() || b.is_ascii_digit() || b == b'-')
}

/// Check if the host is the platform host.
///
/// Platform hosts serve the shared DB (auth, billing, dashboard).
/// Project hosts ({project}.solobase.dev) serve per-project data.
/// localhost/127.0.0.1 only count as platform in dev mode.
pub fn is_platform_host(host: &str, is_dev: bool) -> bool {
    let host_no_port = host.split(':').next().unwrap_or(host);
    if is_dev && (host_no_port == "localhost" || host_no_port == "127.0.0.1") {
        return true;
    }
    host_no_port == "cloud.solobase.dev"
        || host_no_port == "cloud.solobase-dev.dev"
}

/// Per-project configuration stored in the platform D1 database.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectConfig {
    /// Unique project identifier.
    pub id: String,
    /// Subdomain (e.g., `myapp` → `myapp.solobase.dev`).
    pub subdomain: String,
    /// Display name of the project.
    #[serde(default)]
    pub name: String,
    /// Billing plan: "free", "starter", "pro".
    #[serde(default = "default_plan")]
    pub plan: String,
    /// Project status: "active" or "inactive".
    #[serde(default = "default_status")]
    pub status: String,
    /// The user who owns this project.
    #[serde(default)]
    pub owner_user_id: Option<String>,
    /// D1 database UUID for this project.
    #[serde(default)]
    pub db_id: Option<String>,
    /// Whether this is a platform project (has FEATURE_PROJECTS enabled).
    #[serde(default)]
    pub platform: bool,
    /// Custom WASM block names installed by this project.
    #[serde(default)]
    pub blocks: Vec<String>,
}

fn default_plan() -> String { "free".to_string() }
fn default_status() -> String { "active".to_string() }

/// Build a ProjectConfig from a D1 row returned as serde_json::Value.
pub fn project_from_row(row: &serde_json::Value) -> Option<ProjectConfig> {
    Some(ProjectConfig {
        id: row.get("id")?.as_str()?.to_string(),
        subdomain: row.get("subdomain")?.as_str()?.to_string(),
        name: row.get("name").and_then(|v| v.as_str()).unwrap_or("").to_string(),
        plan: row.get("plan").and_then(|v| v.as_str()).unwrap_or("free").to_string(),
        status: row.get("status").and_then(|v| v.as_str()).unwrap_or("active").to_string(),
        owner_user_id: row.get("owner_user_id")
            .and_then(|v| v.as_str())
            .filter(|s| !s.is_empty())
            .map(String::from),
        db_id: row.get("db_id")
            .and_then(|v| v.as_str())
            .filter(|s| !s.is_empty())
            .map(String::from),
        platform: row.get("platform").and_then(|v| v.as_i64()).unwrap_or(0) != 0,
        blocks: Vec::new(),
    })
}

// ---------------------------------------------------------------------------
// Plan limits
// ---------------------------------------------------------------------------

/// Re-export plan limits from the shared crate (single source of truth).
pub use solobase_plans::PlanLimits;

/// Get limits for a plan name.
pub fn get_plan_limits(plan: &str) -> PlanLimits {
    solobase_plans::get_limits(plan)
}

// ---------------------------------------------------------------------------
// Project resolution
// ---------------------------------------------------------------------------

/// Resolve a project from hostname via D1 lookup.
pub async fn resolve_project(
    hostname: &str,
    db: &D1Database,
    is_dev: bool,
) -> std::result::Result<ProjectConfig, String> {
    let host_no_port = hostname.split(':').next().unwrap_or(hostname);
    let subdomain = host_no_port
        .split('.')
        .next()
        .ok_or_else(|| "invalid hostname".to_string())?;

    // Dev mode: return a dev project for localhost
    if is_dev && (subdomain == "localhost" || subdomain == "127") {
        return Ok(ProjectConfig {
            id: "dev".to_string(),
            subdomain: "localhost".to_string(),
            name: "Development".to_string(),
            plan: "pro".to_string(),
            status: "active".to_string(),
            owner_user_id: None,
            db_id: None,
            platform: true,
            blocks: Vec::new(),
        });
    }

    if is_reserved_subdomain(subdomain) {
        return Err(format!("subdomain '{}' is reserved", subdomain));
    }

    let row = db
        .prepare("SELECT * FROM projects WHERE subdomain = ?1")
        .bind(&[subdomain.into()])
        .map_err(|e| format!("D1 bind error: {e}"))?
        .first::<serde_json::Value>(None)
        .await
        .map_err(|e| format!("D1 query error: {e}"))?
        .ok_or_else(|| format!("project '{}' not found", subdomain))?;

    project_from_row(&row)
        .ok_or_else(|| format!("failed to parse project '{}' from D1", subdomain))
}
